projects/plugins/jetpack/_inc/client/components/contextualized-connection/index.jsx

```diff
@@ -0,0 +1,189 @@
+/**
+ * External dependencies
+ */
+import React, { useCallback } from 'react';
+import PropTypes from 'prop-types';
+import { JetpackLogo } from '@automattic/jetpack-components';
+import { __ } from '@wordpress/i18n';
+import { Button } from '@wordpress/components';
+import { createInterpolateElement } from '@wordpress/element';
+import { ConnectButton, ToS } from '@automattic/jetpack-connection';
+
+/**
+ * Internal dependencies
+ */
+import './style.scss';
+
+/**
+ * The Contextualized Connection component.
+ *
+ * @param {object} props -- The properties.
+ * @returns {React.Component} The `ContextualizedConnection` component.
+ */
+const ContextualizedConnection = props => {
+	const {
+		title,
+		logo,
+		children,
+		className,
+		buttonLabel,
+		apiRoot,
+		apiNonce,
+		registrationNonce,
+		redirectUri,
+		autoTrigger,
+		isSiteConnected,
+		setHasSeenWCConnectionModal,
+	} = props;
+
+	const onConnectionRegistered = useCallback( () => {
+		setHasSeenWCConnectionModal( { immediately: false } );
+	}, [ setHasSeenWCConnectionModal ] );
+
+	return (
+		<div className={ 'jp-contextualized-connection' + ( className ? ' ' + className : '' ) }>
+			<div className="jp-contextualized-connection__content">
+				<div className="jp-contextualized-connection__logo">{ logo || <JetpackLogo /> }</div>
+
+				<h2>{ title }</h2>
+
+				{ children }
+
+				{ isSiteConnected && (
+					<Button
+						isPrimary
+						className="jp-contextualized-connection__button"
+						label={ __( 'Continue to Jetpack', 'jetpack' ) }
+						onClick={ setHasSeenWCConnectionModal }
+						href="#/dashboard"
+					>
+						{ __( 'Continue to Jetpack', 'jetpack' ) }
+					</Button>
+				) }
+
+				{ ! isSiteConnected && (
+					<>
+						<ConnectButton
+							autoTrigger={ autoTrigger }
+							apiRoot={ apiRoot }
+							apiNonce={ apiNonce }
+							registrationNonce={ registrationNonce }
+							redirectUri={ redirectUri }
+							connectLabel={ buttonLabel }
+							onRegistered={ onConnectionRegistered }
```

Comment:
I'm afraid `onRegistered` property was removed from `ConnectButton` in #21732 
Thinking of alternatives:
Maybe we could call `setHasSeenWCConnectionModal` on render using `useEffect`. It might actually be better suited there contextually given the user actually sees this component when it's rendered, not when the site is registered.
The downside of this approach (and the reason I'm guessing you didn't choose it in the first place) is that if an error occurs or the user refreshes the page before clicking any buttons here, they won't see the modal again.

Reply:
Yep. Also, it was relying on the state in `main.jsx` to show the component when the Dashboard was visited. That way, if I'd moved it to `useEffect` it would immediately fire a rerender, and it would show another component instead.
But I can do what you said and move the logic I mentioned above to `componentDidMount` in `main.jsx`, then redirect the user to the new registered route (`/woo-step`) if the condition is true. What do you think about that?